src/updater/updater.cpp

```diff
@@ -0,0 +1,640 @@
+#include "updater/updater.hpp"
+
+#include "updater/csv_source.hpp"
+
+#include "extractor/compressed_edge_container.hpp"
+#include "extractor/edge_based_graph_factory.hpp"
+#include "extractor/io.hpp"
+#include "extractor/node_based_edge.hpp"
+
+#include "storage/io.hpp"
+#include "util/exception.hpp"
+#include "util/exception_utils.hpp"
+#include "util/graph_loader.hpp"
+#include "util/integer_range.hpp"
+#include "util/io.hpp"
+#include "util/log.hpp"
+#include "util/static_graph.hpp"
+#include "util/static_rtree.hpp"
+#include "util/string_util.hpp"
+#include "util/timing_util.hpp"
+#include "util/typedefs.hpp"
+
+#include <boost/assert.hpp>
+#include <boost/filesystem/fstream.hpp>
+#include <boost/functional/hash.hpp>
+#include <boost/interprocess/file_mapping.hpp>
+#include <boost/interprocess/mapped_region.hpp>
+
+#include <tbb/blocked_range.h>
+#include <tbb/concurrent_unordered_map.h>
+#include <tbb/enumerable_thread_specific.h>
+#include <tbb/parallel_for_each.h>
+#include <tbb/parallel_invoke.h>
+
+#include <algorithm>
+#include <bitset>
+#include <cstdint>
+#include <fstream>
+#include <iterator>
+#include <memory>
+#include <thread>
+#include <tuple>
+#include <vector>
+
+namespace
+{
+
+template <typename T> inline bool is_aligned(const void *pointer)
+{
+    static_assert(sizeof(T) % alignof(T) == 0, "pointer can not be used as an array pointer");
+    return reinterpret_cast<uintptr_t>(pointer) % alignof(T) == 0;
+}
+
+} // anon ns
+
+namespace osrm
+{
+namespace updater
+{
+
+// Returns duration in deci-seconds
+inline EdgeWeight ConvertToDuration(double distance_in_meters, double speed_in_kmh)
+{
+    if (speed_in_kmh <= 0.)
+        return MAXIMAL_EDGE_DURATION;
+
+    const double speed_in_ms = speed_in_kmh / 3.6;
+    const double duration = distance_in_meters / speed_in_ms;
+    return std::max<EdgeWeight>(1, static_cast<EdgeWeight>(std::round(duration * 10.)));
+}
+
+inline EdgeWeight ConvertToWeight(double weight, double weight_multiplier, EdgeWeight duration)
+{
+    if (std::isfinite(weight))
+        return std::round(weight * weight_multiplier);
+
+    return duration == MAXIMAL_EDGE_DURATION ? INVALID_EDGE_WEIGHT
```

Comment:
We have the problem however that we depend on this functionality right now on our traffic processing. This is not a real problem for us right, since our weights are duration based still anyway.

We should re-think the whole approach on how we ingest weight/duration data though. It is much too easy to screw this up. Can you ticket this?